network/p2p/scoring/registry_test.go

```diff
@@ -422,6 +423,71 @@ func TestPersistingInvalidSubscriptionPenalty(t *testing.T) {
 }
 
+// TestSpamRecordDecayAdjustment ensures that spam record decay is increased each time a peers score reaches the scoring.IncreaseDecayThreshold eventually
+// sustained misbehavior will result in the spam record decay reaching the minimum decay speed .99, and the decay speed is reset to the max decay speed .8.
+func TestSpamRecordDecayAdjustment(t *testing.T) {
+	flowConfig, err := config.DefaultConfig()
+	require.NoError(t, err)
+	scoringRegistryConfig := flowConfig.NetworkConfig.GossipSubConfig.GossipSubScoringRegistryConfig
+	// increase configured DecayRateDecrement so that the decay time is increased faster
+	scoringRegistryConfig.DecayRateDecrement = .1
+	scoringRegistryConfig.DecayAdjustInterval = time.Second
+
+	peer1 := unittest.PeerIdFixture(t)
+	peer2 := unittest.PeerIdFixture(t)
+	reg, spamRecords := newScoringRegistry(
+		t,
+		scoringRegistryConfig,
+		withStakedIdentity(peer1),
+		withValidSubscriptions(peer1),
+		withStakedIdentity(peer2),
+		withValidSubscriptions(peer2))
+
+	// initially, the spamRecords should not have the peer ids.
+	assert.False(t, spamRecords.Has(peer1))
+	assert.False(t, spamRecords.Has(peer2))
+
+	// simulate sustained malicious activity from peer1, eventually the decay speed
+	// for a spam record should be reduced to the MinimumSpamPenaltyDecaySpeed
+	require.Eventually(t, func() bool {
+		reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+			PeerID:  peer1,
+			MsgType: p2pmsg.CtrlMsgPrune,
+		})
+		record, err, ok := spamRecords.Get(peer1)
+		require.NoError(t, err)
+		require.True(t, ok)
+		return record.Decay == scoring.MinimumSpamPenaltyDecaySpeed
+	}, 5*time.Second, 500*time.Millisecond)
+
+	// initialize a spam record for peer2
+	reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+		PeerID:  peer2,
+		MsgType: p2pmsg.CtrlMsgPrune,
+	})
+	// reduce penalty and increase Decay to scoring.MinimumSpamPenaltyDecaySpeed
+	record, err := spamRecords.Update(peer2, func(record p2p.GossipSubSpamRecord) p2p.GossipSubSpamRecord {
+		record.Penalty = -.1
+		record.Decay = scoring.MinimumSpamPenaltyDecaySpeed
+		return record
+	})
+	require.NoError(t, err)
+	require.True(t, record.Decay == scoring.MinimumSpamPenaltyDecaySpeed)
+	require.True(t, record.Penalty == -.1)
+	// simulate sustained good behavior from peer 2, each time the spam record is read from the cache
+	// using Get method the record penalty will be decayed until it is eventually reset to
+	// 0 at this point the decay speed for the record should be reset to MaximumSpamPenaltyDecaySpeed
+	// eventually after penalty reaches the skipDecaThreshold the record decay will be reset to scoring.MaximumSpamPenaltyDecaySpeed
+	require.Eventually(t, func() bool {
+		record, err, ok := spamRecords.Get(peer2)
+		require.NoError(t, err)
+		require.True(t, ok)
+		return record.Decay == scoring.MaximumSpamPenaltyDecaySpeed &&
+			record.Penalty == 0 &&
+			record.LastDecayAdjustment.IsZero()
+	}, 5*time.Second, time.Second)
```

Comment:
After this line, please add a persisting spam attack from peer2 (similar to what you have already done for peer1) to ensure that decay speed can be reduced again back to the minimum factor. This evaluates the end-to-end health of the initialization -> decay -> restore -> decay cycle.

Reply:
https://github.com/onflow/flow-go/pull/4891/commits/8885c1c0002a7ad64e79d6d124c5dfa3de269da0